Mathlib/ModelTheory/PartialEquiv.lean

```diff
@@ -269,6 +273,217 @@ theorem toPartialEquiv_toEmbedding {f :  M ≃ₚ[L] N} (h : f.dom = ⊤) :
 
 end Embedding
 
+namespace DirectLimit
+
+open PartialEquiv
+
+variable {ι : Type*} [Preorder ι] [Nonempty ι] [IsDirected ι (· ≤ ·)]
+variable (S : ι →o M ≃ₚ[L] N)
+
+instance : DirectedSystem (fun i ↦ (S i).dom) (fun _ _ h ↦
+  Substructure.inclusion (dom_le_dom (S.monotone h))) where
+  map_self' := fun _ _ _ ↦ rfl
+  map_map' := fun _ _ _ ↦ rfl
+
+instance : DirectedSystem (fun i ↦ (S i).cod) (fun _ _ h ↦
+  Substructure.inclusion (cod_le_cod (S.monotone h))) where
+  map_self' := fun _ _ _ ↦ rfl
+  map_map' := fun _ _ _ ↦ rfl
+
+/-- The limit of a directed system of PartialEquivs.-/
+noncomputable def partialEquivLimit : M ≃ₚ[L] N where
+  dom := iSup (fun i ↦ (S i).dom)
+  cod := iSup (fun i ↦ (S i).cod)
+  toEquiv :=
+    (Equiv_iSup {
+      toFun := (fun i ↦ (S i).cod)
+      monotone' := monotone_cod.comp S.monotone}
+    ).comp
+      ((DirectLimit.equiv_lift L ι (fun i ↦ (S i).dom)
+        (fun _ _ hij ↦ Substructure.inclusion (dom_le_dom (S.monotone hij)))
+        (fun i ↦ (S i).cod)
+        (fun _ _ hij ↦ Substructure.inclusion (cod_le_cod (S.monotone hij)))
+        (fun i ↦ (S i).toEquiv)
+        (fun _ _ hij _ ↦ toEquiv_inclusion_apply (S.monotone hij) _)
+      ).comp
+        (Equiv_iSup {
+          toFun := (fun i ↦ (S i).dom)
+          monotone' := monotone_dom.comp S.monotone}).symm)
+
+@[simp]
+theorem partialEquiv_limit_dom : (partialEquivLimit S).dom = iSup (fun x ↦ (S x).dom) := rfl
+
+@[simp]
+theorem partialEquiv_limit_cod : (partialEquivLimit S).cod = iSup (fun x ↦ (S x).cod) := rfl
+
+@[simp]
+lemma partialEquivLimit_comp_inclusion {i : ι} :
+    (partialEquivLimit S).toEquiv.toEmbedding.comp (Substructure.inclusion (le_iSup _ i)) =
+    (Substructure.inclusion (le_iSup _ i)).comp (S i).toEquiv.toEmbedding := by
+  simp only [partialEquivLimit, Equiv.comp_toEmbedding, Embedding.comp_assoc]
+  rw [Equiv_iSup_symm_inclusion]
+  congr
+
+theorem le_partialEquivLimit : ∀ i, S i ≤ partialEquivLimit S :=
+  fun i => ⟨le_iSup (f := fun i ↦ (S i).dom) _, by
+    simp only [partialEquiv_limit_cod, partialEquiv_limit_dom, partialEquivLimit_comp_inclusion, ←
+      Embedding.comp_assoc, subtype_comp_inclusion]⟩
+
+end DirectLimit
+
+section FGPartialEquiv
+
+open PartialEquiv Set DirectLimit
+
+variable (M) (N) (L)
+
+/-- The type of equivalences between finitely generated substructures. -/
+abbrev FiniteEquiv := {f : M ≃ₚ[L] N // f.dom.FG}
+
+variable {M} {N} {L}
+
+/-- The cofinal set of finite equivalences with a given element in their domain. -/
+def definedAtLeft
+    (h : ∀ f : (M ≃ₚ[L] N), ∀ _ : f.dom.FG, ∀ m : M, ∃ g : (M ≃ₚ[L] N), f ≤ g ∧ m ∈ g.dom)
+    (m : M) : Order.Cofinal (FiniteEquiv L M N) where
```

Comment:
Are you sure you want to use `Order.Cofinal` here rather than a `Set (FiniteEquiv L M N)` + `∃ᶠ f in atTop, f ∈ mySet`? or even maybe just the statement `∃ᶠ f in atTop, m ∈ f.val.dom`?

Reply:
🏓 You seem to have let this through

Reply:
I am now sure that I want to keep `Order.Cofinal` instead of using filters, because `L.FGEquiv M N` isn't directed, so `(atTop : Filter (L.FGEquiv M N)) = ⊥`.
Specifically, look at two partial equivalences sending a particular element to two different values - there can't be a common extension.

Reply:
Oh so `∀ a, ∃ b ≥ a, p b` doesn't even describe a filter. That's unfortunate yes